File: android/src/main/java/com/visibilityawareview/VisibilityAwareView.java
package com.visibilityawareview;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import android.content.Context;
import android.graphics.Rect;
import android.os.Handler;
import android.util.AttributeSet;

import android.view.View;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.UIManagerHelper;
import com.facebook.react.uimanager.UIManagerModule;
import com.facebook.react.uimanager.events.Event;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.facebook.react.views.view.ReactViewGroup;

public class VisibilityAwareView extends ReactViewGroup implements View.OnAttachStateChangeListener, View.OnLayoutChangeListener, LifecycleEventListener {
  private static final long TIMER_ACCURACY = 250;

  private boolean isAppInBackground = false;
  private Boolean _attachedToWindow = false;

  private long _interval_accuracy;
  private double _minVisibleArea;
  private Handler _intervalHandler;
  private boolean _isVisible;
  private boolean _ignoreAppState;

  protected ReactApplicationContext reactApplicationContext;

  /**
   * Getters
   */

  public Boolean isViewVisible() {
    return _isVisible;
  }

  /**
   * Setters
   */

  public VisibilityAwareView(ThemedReactContext context, @Nullable ReactApplicationContext reactApplicationContext) {
    super(context);

    if(reactApplicationContext != null) {
      this.reactApplicationContext = reactApplicationContext;
      this.reactApplicationContext.addLifecycleEventListener(this);
    }
    _interval_accuracy = TIMER_ACCURACY;
    _minVisibleArea = 0.01;
    _ignoreAppState = false;

    this.addOnLayoutChangeListener(this);
    this.addOnAttachStateChangeListener(this);

    this.updateVisibilityIfNeeded();
  }

  public void setTrackingAccuracy(@Nullable double accuracy) {
    if (accuracy > 0) {
      _interval_accuracy = (long) accuracy;
    }
  }

  public void  setMinVisibleArea(@Nullable double minVisibleArea) {
    _minVisibleArea = minVisibleArea;
  }

  public void  setIgnoreAppState(@Nullable boolean ignoreAppState) {
    _ignoreAppState = ignoreAppState;
  }

  public void start() {
    this.startIntervalTrackingIfOnWindow();
  }

  public void stop(){
    this.stopIntervalForVisibilityCheck();
  }


  /**
   * Update / Check Visibility
   */

  private Rect getRectForViewInWindow(View v) {
    int[] locationOnWindow = new int[2];
    v.getLocationInWindow(locationOnWindow);
    Rect r = new Rect();
    v.getDrawingRect(r);
    int x = locationOnWindow[0];
    int y = locationOnWindow[1];
    int w = v.getMeasuredWidth();
    int h = v.getMeasuredHeight();
    return new Rect(x, y, x + w, y + h);
  }

  private boolean isViewVisibleWithinView(View view, View inView, int depth) {
    if (inView == null) {
      return true;
    }

    Rect viewR = this.getRectForViewInWindow(view);
    Rect inR = this.getRectForViewInWindow(inView);

    Boolean intersects = viewR.intersect(inR);
    if(intersects){
      Rect overlappingRect = new Rect();
      overlappingRect.setIntersect(viewR, inR);
      int overlappingArea = overlappingRect.width() * overlappingRect.height();
      int totalArea = view.getMeasuredWidth() * view.getMeasuredHeight();
      int totalArea_viewR = viewR.width() * viewR.height();
      float overlapPercentage = (float) overlappingArea / totalArea;
      if (overlapPercentage < _minVisibleArea) {
        return false;
      }
    }
    if (intersects) {
      if (inView.getParent() instanceof View) {
        return this.isViewVisibleWithinView(view, (View) inView.getParent(), depth + 1);
      } else {
        return true;
      }
    }
    return false;
  }

  private Boolean isViewVisibleWithinScreen() {
    Boolean hasWindow = getWindowToken() != null;
    if (!hasWindow) {
      return false;
    }

    boolean visible = this.isViewVisibleWithinView(this, (View) this.getParent(), 0);
    return visible;
  }

  public void updateVisibilityIfNeeded() {
    Boolean attached = this._attachedToWindow;
    Boolean frameVisible = attached && this.isViewVisibleWithinScreen();
    Boolean nextState;
    if(_ignoreAppState){
      nextState = attached && frameVisible;
    }else {
      nextState = !this.isAppInBackground && attached && frameVisible;
    }
    if (this.isViewVisible() != nextState) {
      this.updateVisibility(nextState, frameVisible);
    }
  }

  private void updateVisibility(Boolean nextState, Boolean frameVisible) {
    if (this.isViewVisible() != nextState) {
      _isVisible = nextState;
      if (_isVisible) {
        this.startIntervalTrackingIfNotRunningAlready();
      } else {
      }

      this.fireOnChangeVisibilityEvent(frameVisible);
    }
  }


  /**
   * Fire Callbacks + helper
   */

  public VisibilityAwareViewHandler handler;

  interface VisibilityAwareViewHandler {
    void viewDidEnterVisibleArea(VisibilityAwareView customView, WritableMap eventData);
    void viewDidLeaveVisibleArea(VisibilityAwareView customView, WritableMap eventData);
  }

  public static class ChangeVisibilityEvent extends Event<ChangeVisibilityEvent> {
    public static final String BECOME_VISIBLE_EVENT_NAME = "topOnBecomeVisible";
    public static final String BECOME_INVISIBLE_EVENT_NAME = "topOnBecomeInvisible";

    private final boolean isVisible;
    private final WritableMap eventData;

    public ChangeVisibilityEvent(int viewId, boolean isVisible, WritableMap data) {
      super(viewId);
      this.isVisible = isVisible;
      this.eventData = data;
    }

    @Override
    public String getEventName() {
      return this.isVisible ? BECOME_VISIBLE_EVENT_NAME : BECOME_INVISIBLE_EVENT_NAME;
    }

    @Override
    public short getCoalescingKey() {
      return 0;
    }

    @Override
    public boolean canCoalesce() {
      return false;
    }

    @Nullable
    @Override
    protected WritableMap getEventData() {
      return serializeEventData();
    }

    private WritableMap serializeEventData() {
      eventData.putInt("target", getViewTag());
      return eventData;
    }
  }



  private void fireOnChangeVisibilityEvent(Boolean frameVisible) {
    WritableMap eventData = Arguments.createMap();
    eventData.putBoolean("frame_visible", frameVisible);
    if (this.isViewVisible()){
      eventData.putBoolean("app_open", !this.isAppInBackground);
    }else{
      eventData.putBoolean("app_closed", this.isAppInBackground);
    }

    if(this.handler != null){
      if (this.isViewVisible()){
        this.handler.viewDidEnterVisibleArea(this, eventData);
      }else{
        this.handler.viewDidLeaveVisibleArea(this, eventData);
      }
    }

  }

  /**
   * Timer Handling
   */
  private void startIntervalForVisibilityCheck() {

    if (this._intervalHandler != null) {
      this._intervalHandler.removeCallbacksAndMessages(null);
      this._intervalHandler = null;
    }
    this._intervalHandler = new Handler();
    this._intervalHandler.post(new Runnable() {
      @Override
      public void run() {
        updateVisibilityIfNeeded();
        if (_intervalHandler != null) {
          _intervalHandler.postDelayed(this, _interval_accuracy);
        }
      }
    });

  }

  private void startIntervalTrackingIfNotRunningAlready() {
    if (this._intervalHandler == null) {
      this.startIntervalForVisibilityCheck();
    }
  }

  private void startIntervalTrackingIfOnWindow() {
    if (getWindowToken() != null) {
      this.startIntervalTrackingIfNotRunningAlready();
    }
  }

  private void stopIntervalForVisibilityCheck() {
    if (this._intervalHandler != null) {
      this._intervalHandler.removeCallbacksAndMessages(null);
    }
    this._intervalHandler = null;
  }

  /**
   * Functions which should cause a visibility check
   */

  @Override
  public void onViewAttachedToWindow(@NonNull View view) {
    this._attachedToWindow = true;
    this.updateVisibilityIfNeeded();
    this.startIntervalTrackingIfOnWindow();
  }

  @Override
  public void onViewDetachedFromWindow(@NonNull View view) {
    this._attachedToWindow = false;
    this.updateVisibilityIfNeeded();
    this.stopIntervalForVisibilityCheck();
  }

  @Override
  public void onLayoutChange(View view, int i, int i1, int i2, int i3, int i4, int i5, int i6, int i7) {
    this.updateVisibilityIfNeeded();
  }

  @Override
  public void onHostResume() {
    this.isAppInBackground = false;
    this.updateVisibilityIfNeeded();
    this.startIntervalTrackingIfOnWindow();
  }

  @Override
  public void onHostPause() {
    this.isAppInBackground = true;
    this.stopIntervalForVisibilityCheck();
    this.updateVisibilityIfNeeded();
  }

  @Override
  public void onHostDestroy() {
    this.stopIntervalForVisibilityCheck();
    this.removeOnLayoutChangeListener(this);
    this.removeOnAttachStateChangeListener(this);
    if(this.reactApplicationContext != null){
      this.reactApplicationContext.removeLifecycleEventListener(this);
    }
  }
}

File: android/src/main/java/com/visibilityawareview/VisibilityAwareView.kt
package com.visibilityawareview

import android.graphics.Rect
import android.os.Handler
import android.os.Looper
import android.view.View
import com.facebook.react.bridge.*
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.views.view.ReactViewGroup

class VisibilityAwareView(
  context: ThemedReactContext,
  private val reactContext: ReactApplicationContext?
) : ReactViewGroup(context),
  View.OnAttachStateChangeListener,
  View.OnLayoutChangeListener,
  LifecycleEventListener {

  companion object {
    private const val TIMER_ACCURACY = 250L
  }

  private var isAppInBackground = false
  private var attachedToWindow = false

  private var intervalAccuracy = TIMER_ACCURACY
  private var minVisibleArea = 0.01
  private var intervalHandler: Handler? = null
  private var isVisible = false
  private var ignoreAppState = false

  var handler: VisibilityAwareViewHandler? = null

  init {
    reactContext?.addLifecycleEventListener(this)
    addOnLayoutChangeListener(this)
    addOnAttachStateChangeListener(this)
    updateVisibilityIfNeeded()
  }

  fun start() = startIntervalTrackingIfOnWindow()
  fun stop() = stopInterval()

  fun setTrackingAccuracy(value: Double?) {
    if (value != null && value > 0) {
      intervalAccuracy = value.toLong()
    }
  }

  fun setMinVisibleArea(value: Double?) {
    if (value != null) minVisibleArea = value
  }

  fun setIgnoreAppState(value: Boolean?) {
    ignoreAppState = value ?: false
  }

  private fun getRect(view: View): Rect {
    val loc = IntArray(2)
    view.getLocationInWindow(loc)
    return Rect(loc[0], loc[1], loc[0] + view.width, loc[1] + view.height)
  }

  private fun isVisibleInParent(view: View, parent: View?): Boolean {
    parent ?: return true

    val viewRect = getRect(view)
    val parentRect = getRect(parent)

    if (!viewRect.intersect(parentRect)) return false

    val overlap = Rect()
    overlap.setIntersect(viewRect, parentRect)

    val overlapArea = overlap.width() * overlap.height()
    val totalArea = view.width * view.height

    if (totalArea > 0) {
      val percent = overlapArea.toFloat() / totalArea
      if (percent < minVisibleArea) return false
    }

    return if (parent.parent is View) {
      isVisibleInParent(view, parent.parent as View)
    } else true
  }

  private fun isVisibleOnScreen(): Boolean {
    if (windowToken == null) return false
    return isVisibleInParent(this, parent as? View)
  }

  fun updateVisibilityIfNeeded() {
    val frameVisible = attachedToWindow && isVisibleOnScreen()

    val next = if (ignoreAppState) {
      attachedToWindow && frameVisible
    } else {
      !isAppInBackground && attachedToWindow && frameVisible
    }

    if (isVisible != next) {
      isVisible = next
      fireEvent(frameVisible)
    }
  }

  private fun fireEvent(frameVisible: Boolean) {
    val map = Arguments.createMap().apply {
      putBoolean("frame_visible", frameVisible)
      if (isVisible) {
        putBoolean("app_open", !isAppInBackground)
      } else {
        putBoolean("app_closed", isAppInBackground)
      }
    }

    if (isVisible) {
      handler?.viewDidEnterVisibleArea(this, map)
    } else {
      handler?.viewDidLeaveVisibleArea(this, map)
    }
  }

  private fun startIntervalTrackingIfOnWindow() {
    if (windowToken != null && intervalHandler == null) {
      intervalHandler = Handler(Looper.getMainLooper())
      intervalHandler?.post(object : Runnable {
        override fun run() {
          updateVisibilityIfNeeded()
          intervalHandler?.postDelayed(this, intervalAccuracy)
        }
      })
    }
  }

  private fun stopInterval() {
    intervalHandler?.removeCallbacksAndMessages(null)
    intervalHandler = null
  }

  override fun onViewAttachedToWindow(v: View) {
    attachedToWindow = true
    updateVisibilityIfNeeded()
    startIntervalTrackingIfOnWindow()
  }

  override fun onViewDetachedFromWindow(v: View) {
    attachedToWindow = false
    stopInterval()
    updateVisibilityIfNeeded()
  }

  override fun onLayoutChange(
    v: View, l: Int, t: Int, r: Int, b: Int,
    oldL: Int, oldT: Int, oldR: Int, oldB: Int
  ) {
    updateVisibilityIfNeeded()
  }

  override fun onHostResume() {
    isAppInBackground = false
    updateVisibilityIfNeeded()
    startIntervalTrackingIfOnWindow()
  }

  override fun onHostPause() {
    isAppInBackground = true
    stopInterval()
    updateVisibilityIfNeeded()
  }

  override fun onHostDestroy() {
    stopInterval()
    removeOnLayoutChangeListener(this)
    removeOnAttachStateChangeListener(this)
    reactContext?.removeLifecycleEventListener(this)
  }

  interface VisibilityAwareViewHandler {
    fun viewDidEnterVisibleArea(view: VisibilityAwareView, data: WritableMap)
    fun viewDidLeaveVisibleArea(view: VisibilityAwareView, data: WritableMap)
  }
}